[Interactive] Integrate Recommendation capability into `interactive` (#5950)
Azure/azure-cli-extensions: src/interactive/README.md

**Usage**: When you don't know what to do next, you can try entering only `Space` first, it will recommend the next command which is most used in other similar sessions.

<img align=center src="docs/command_recommendation.gif" width = "70%" alt="command_recommendation">

### Scenario Identification

Scenario identification is another powerful feature that automatically recognizes the user's current scenario and recommends a set of commands that are most relevant to the user's needs. This feature is based on our extensive knowledge base which cover a wide range of use cases across various scenes. By analyzing the user's input history, Intelligent Az Interactive can identify the relevant scenario and suggest the most appropriate command set, helping new CLI users to quickly accomplish their tasks with confidence and ease.

**Usage**: You can enter `Space` and select the scenario *(the options with prefix `command set`)* you want to use. It will automatically execute the commands in the scenario and help you complete the task.

<img align=center src="docs/scenario_recommendation.gif" width = "70%" alt="scenario_recommendation">

> ## Search for Scenarios

Reviewer: ```suggestion
## Search Scenario Knowledge
```

Additionally, the loading bar helps prevent user interface lagging or freezing due to insufficient memory. By preloading the necessary resources and data, users can avoid encountering these issues when inputting commands. The loading bar is an essential tool that improves the user experience by ensuring that all necessary components have been loaded before usage, thereby reducing the risk of encountering errors and providing a smoother, more stable platform.

If you want to disable the loading bar, you can run the following command:
```bash
$ az config set enable_preloading=False
```
Please refer to the following gif to see the loading bar in action:

<img src="docs/loading_bar.gif" width = "70%" alt="loading_bar" align=center>

## Value Completion in Scenarios

> We added a completion mechanism for parameter values in scenarios to improve the completion ability of param value in
> scenarios. In multiple commands of the same scenario, once the user enters a parameter value, we store the value entered by

Reviewer: ```suggestion
We added a completion mechanism for parameter values in scenario execution *(corresponding to scenario identification or searching scenario knowledge)* to improve the completion ability of param value in
scenarios. In multiple commands of the same scenario, once the user enters a parameter value, we store the value entered by
```

```

The color option will be saved.

# Intelligent AZ Interactive (Revolutionary Change)
## Recommendation

We have integrated the cli recommendation to make the completion ability more intelligent and provide the scenario completion. This is a revolutionary change. Users can enable or disable the recommendation feature by running the following commands:

```

> $ az config set interactive.enable_recommender=True # Default, try the new recommendation feature

Reviewer 2: Is it enabled by default?

The command recommendation feature can also be a good way to broaden users' knowledge boundaries and help them recognize and learn commands that are related to their common commands but never used before.

### Scenario Identification

Scenario identification is another powerful feature that automatically recognizes the user's current scenario and recommends a set of commands that are most relevant to the user's needs. This feature is based on our extensive database of over 600+ business scenarios, which cover a wide range of use cases across various scenes. By analyzing the user's input history, Intelligent Az Interactive can identify the relevant scenario and suggest the most appropriate command set, helping new CLI users to quickly accomplish their tasks with confidence and ease.

## Search for Scenarios

> We have added the ability to help users search for scenarios based on keywords. When the user enters some keywords or descriptions, we will recommend the scenarios that are most likely to be used based on the keywords and descriptions of the functions the user wants to implement.

Reviewer: We need to inform customers in the `README.md` how to use or trigger these new features

Author: Thank you for the suggestion, already modified Please refer to: https://github.com/wu5bocheng/azure-cli-extensions/blob/interactive-main/src/interactive/README.md

```
$ az interactive // initialize the az interactive
$ /connect a mongodb to web app // Search for scenario by starting with / and entering keywords
>>  output
[1] Connect an app to MongoDB (Cosmos DB). (5 Commands)
Connect an app to MongoDB (Cosmos DB).

[2] Tutorial to create and connect Web App to Azure Database for MySQL Flexible Server in a virtual network (6 Commands)
Tutorial to create and connect Web App to Azure Database for MySQL Flexible Server in a virtual network

[3] Connect an app to SQL Database. (7 Commands)
Connect an app to SQL Database.

[4] Connect an app to a storage account. (5 Commands)
Connect an app to a storage account.

[5] Deploy an ASP.NET Core web app to Azure App Service and connect to an Azure SQL Database. (8 Commands)
Deploy an ASP.NET Core web app to Azure App Service and connect to an Azure SQL Database.

 ? Please select your option (if none, enter 0):
 $ 1 // Select the scenario you want to use
```

## Loading Bar

> The loading bar is a feature that we have implemented to provide users with a more stable and predictable experience while using our platform. Its primary function is to prevent command parameter detection errors that may occur due to incomplete loading. By displaying the progress of initialization through the loading bar, users can be assured that the initialization process is ongoing and they can expect to see a fully loaded interface once the bar reaches 100%.

Reviewer 2: - How long will it usually take to load all commands/parameters?
- Do we have a configure to be able to disable the loading bar if customers don't want it?
- Is it enabled by default?

Author: Our loading bar feature is automatically enabled, and users can turn it off by `az config set enable_preloading=False`. The average time for preloading is 30-40s, depending on the machine's memory strength.
Our preloading feature does not increase the loading time compared to the original case, but it will give the user the perception of increased loading time. Although preloading will avoid many command run errors due to incomplete command loading, further optimization of loading time can be done in the future.

Additionally, the loading bar helps prevent user interface lagging or freezing due to insufficient memory. By preloading the necessary resources and data, users can avoid encountering these issues when inputting commands.

> The loading bar is an essential tool that improves the user experience by ensuring that all necessary components are loaded before usage, thereby reducing the risk of encountering errors and providing a smoother, more stable platform.

Reviewer 3: ```suggestion
The loading bar is an essential tool that improves the user experience by ensuring that all necessary components have been loaded before usage, thereby reducing the risk of encountering errors and providing a smoother, more stable platform.
```

```
Here is a demo video demonstrating the main features of the new az interactive:

[![intelligent_az_interactive](docs/intelligent_az_interactive.png)](docs/intelligent_az_interactive.mp4 "intelligent_az_interactive")

### Command Recommendation

Command recommendation uses an intelligent algorithm to suggest the most relevant and frequently used command for the user based on their historical command inputs. By analyzing the user's input history and matching it with the highest relevant commands, Intelligent Az Interactive can assist Azure CLI beginners in making quick and informed decisions, saving them time and effort.

The command recommendation feature can also be a good way to broaden users' knowledge boundaries and help them recognize and learn commands that are related to their common commands but never used before.

When you don't know what to do next, you can enter Space, and it will recommend the next command which is most used in other similar sessions.

<img align=center src="docs/command_recommendation.gif" width = "70%" alt="command_recommendation">

### Scenario Identification

> Scenario identification is another powerful feature that automatically recognizes the user's current scenario and recommends a set of commands that are most relevant to the user's needs. This feature is based on our extensive database of over 600+ business scenarios, which cover a wide range of use cases across various scenes. By analyzing the user's input history, Intelligent Az Interactive can identify the relevant scenario and suggest the most appropriate command set, helping new CLI users to quickly accomplish their tasks with confidence and ease.

Reviewer: ```suggestion
Scenario identification is another powerful feature that automatically recognizes the user's current scenario and recommends a set of commands that are most relevant to the user's needs. This feature is based on our extensive knowledge base which cover a wide range of use cases across various scenes. By analyzing the user's input history, Intelligent Az Interactive can identify the relevant scenario and suggest the most appropriate command set, helping new CLI users to quickly accomplish their tasks with confidence and ease.
```

> You can enter Space and select the scenario you want to use. It will automatically execute the commands in the scenario and help you complete the task.

Reviewer: ```suggestion
**Usage**: You can enter `Space` and select the scenario *(the options with prefix `command set`)* you want to use. It will automatically execute the commands in the scenario and help you complete the task.
```

<img align=center src="docs/scenario_recommendation.gif" width = "70%" alt="scenario_recommendation">

## Search for Scenarios

We have added the ability to help users search for scenarios based on keywords. When the user enters some keywords or descriptions, we will recommend the scenarios that are most likely to be used based on the keywords and descriptions of the functions the user wants to implement.

> You can use the natural language keywords with prefix "/" to search for the commands and usage scenarios you need, and smoothly create and execute corresponding CLI tasks.

Reviewer: ```suggestion
**Usage**: You can use the natural language keywords with prefix `/` to search for the commands and usage scenarios you need, and smoothly create and execute corresponding CLI tasks.
```

<img align=center src="docs/scenario_search.gif" width = "70%" alt="scenario_search">

> ```
> $ az interactive // initialize the az interactive
> $ /connect a mongodb to web app // Search for scenario by starting with / and entering keywords
> >>  output
> [1] Connect an app to MongoDB (Cosmos DB). (5 Commands)
> Connect an app to MongoDB (Cosmos DB).
> 
> [2] Tutorial to create and connect Web App to Azure Database for MySQL Flexible Server in a virtual network (6 Commands)
> Tutorial to create and connect Web App to Azure Database for MySQL Flexible Server in a virtual network
> 
> [3] Connect an app to SQL Database. (7 Commands)
> Connect an app to SQL Database.
> 
> [4] Connect an app to a storage account. (5 Commands)
> Connect an app to a storage account.
> 
> [5] Deploy an ASP.NET Core web app to Azure App Service and connect to an Azure SQL Database. (8 Commands)
> Deploy an ASP.NET Core web app to Azure App Service and connect to an Azure SQL Database.
> 
>  ? Please select your option (if none, enter 0):
>  $ 1 // Select the scenario you want to use
> ```

Reviewer: ```suggestion
```
```
```

Additionally, the loading bar helps prevent user interface lagging or freezing due to insufficient memory. By preloading the necessary resources and data, users can avoid encountering these issues when inputting commands. The loading bar is an essential tool that improves the user experience by ensuring that all necessary components have been loaded before usage, thereby reducing the risk of encountering errors and providing a smoother, more stable platform.

If you want to disable the loading bar, you can run the following command:
```bash
$ az config set enable_preloading=False
```
Please refer to the following gif to see the loading bar in action:

<img src="docs/loading_bar.gif" width = "70%" alt="loading_bar" align=center>

## Value Completion in Scenarios

We added a completion mechanism for parameter values in scenarios to improve the completion ability of param value in
scenarios. In multiple commands of the same scenario, once the user enters a parameter value, we store the value entered by
the user based on the scenario sample value and some special global parameters, and automatically recommend the completion
of these parameter values in subsequent commands.

> ## Telemetry Feedback Optimization
> 
> To collect data and facilitate the optimization and tuning of the cli recommendation model, we have optimized the
> telemetry feedback function. We have added `CLIRecommendation` to the `properties` of telemetry feedback. For details,
> please refer
> to [cli-recommendation](https://github.com/hackathon-cli-recommendation/cli-recommendation/blob/master/Docs/feedback_design.md).

Reviewer: For customers, they don't need to know this
```suggestion
```
